In [36]:
# Portfolio Project: Online Retail Exploratory Data Analysis with Python

## Overview

In this project, you will step into the shoes of an entry-level data analyst at an online retail company, helping interpret real-world data to help make a key business decision.

Source files can be found here: https://www.coursera.org/learn/perform-exploratory-data-analysis-on-retail-data-with-python/home/week/1

## Case Study
In this project, you will be working with transactional data from an online retail store. The dataset contains information about customer purchases, including product details, quantities, prices, and timestamps. Your task is to explore and analyze this dataset to gain insights into the store's sales trends, customer behavior, and popular products. 

By conducting exploratory data analysis, you will identify patterns, outliers, and correlations in the data, allowing you to make data-driven decisions and recommendations to optimize the store's operations and improve customer satisfaction. Through visualizations and statistical analysis, you will uncover key trends, such as the busiest sales months, best-selling products, and the store's most valuable customers. Ultimately, this project aims to provide actionable insights that can drive strategic business decisions and enhance the store's overall performance in the competitive online retail market.

## Project Objectives
1. Describe data to answer key questions to uncover insights
2. Gain valuable insights that will help improve online retail performance
3. Provide analytic insights and data-driven recommendations

## Dataset

The dataset you will be working with is the "Online Retail" dataset. It contains transactional data of an online retail store from 2010 to 2011. The dataset is available as a .xlsx file named `Online Retail.xlsx`. This data file is already included in the Coursera Jupyter Notebook environment, however if you are working off-platform it can also be downloaded [here](https://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx).

The dataset contains the following columns:

- InvoiceNo: Invoice number of the transaction
- StockCode: Unique code of the product
- Description: Description of the product
- Quantity: Quantity of the product in the transaction
- InvoiceDate: Date and time of the transaction
- UnitPrice: Unit price of the product
- CustomerID: Unique identifier of the customer
- Country: Country where the transaction occurred

## Tasks

You may explore this dataset in any way you would like - however if you'd like some help getting started, here are a few ideas:

1. Load the dataset into a Pandas DataFrame and display the first few rows to get an overview of the data.
2. Perform data cleaning by handling missing values, if any, and removing any redundant or unnecessary columns.
3. Explore the basic statistics of the dataset, including measures of central tendency and dispersion.
4. Perform data visualization to gain insights into the dataset. Generate appropriate plots, such as histograms, scatter plots, or bar plots, to visualize different aspects of the data.
5. Analyze the sales trends over time. Identify the busiest months and days of the week in terms of sales.
6. Explore the top-selling products and countries based on the quantity sold.
7. Identify any outliers or anomalies in the dataset and discuss their potential impact on the analysis.
8. Draw conclusions and summarize your findings from the exploratory data analysis.

# Task 1: Load the Data

In [37]:
import pandas as pd
import sqlalchemy
import psycopg2

pd.set_option('display.max_rows', 50)
pd.set_option('display.max_columns', 15)
pd.set_option('expand_frame_repr', False)

In [38]:
# %%timeit
# excel_data_df = pd.read_excel('Online Retail.xlsx', sheet_name='Online Retail')

> 48.2 s ± 600 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

It takes too long (over 6 minutes) to read an Excel file of this size. Instead, I loaded the data into a local Postgres database so it can be read in 1 or 2 seconds.

In [39]:
engine = sqlalchemy.create_engine('postgresql://postgres:password@localhost:5432/eda_retail_data')
retail_data = pd.read_sql_table(
    "retail_data",
    con=engine,
    columns=['InvoiceNo',
             'StockCode',
             'Description',
             'Quantity',
             'InvoiceDate',
             'UnitPrice',
             'CustomerId',
             'Country'])

In [40]:
retail_data.sample(50)

InvoiceNo StockCode                          Description  Quantity InvoiceDate  UnitPrice  CustomerId         Country
276499    561051     23203            JUMBO BAG DOILEY PATTERNS       300  2011-07-24       1.79     16684.0  United Kingdom
313333    564483     22332       SKULLS PARTY BAG + STICKER SET         8  2011-08-25       1.65     14050.0  United Kingdom
245089    558571     22556       PLASTERS IN TIN CIRCUS PARADE         12  2011-06-30       1.65     17017.0  United Kingdom
38512     539580     22511          RETROSPOT BABUSHKA DOORSTOP         4  2010-12-20       3.75     15312.0  United Kingdom
238884    557958     21559    STRAWBERRY LUNCH BOX WITH CUTLERY         4  2011-06-23       2.55     16161.0  United Kingdom
176078    551995     20707         CRAZY DAISY HEART DECORATION         2  2011-05-05       0.83         NaN  United Kingdom
408599    571965     22988                    SOLDIERS EGG CUP          2  2011-10-20       1.25     17238.0  United Kingdom
528754    580729    85169A               IVORY LOVE BIRD CANDLE         2  2011-12-05       2.46         NaN  United Kingdom
505611    579093     35953   FOLKART STAR CHRISTMAS DECORATIONS        48  2011-11-28       0.39     13081.0  United Kingdom
304853    563613     21669      BLUE STRIPE CERAMIC DRAWER KNOB        96  2011-08-18       1.06     12688.0          Israel
450206    575176     16237                 SLEEPING CAT ERASERS         7  2011-11-08       0.42         NaN  United Kingdom
78626     542891    82494L          WOODEN FRAME ANTIQUE WHITE         48  2011-02-01       2.55     16705.0  United Kingdom
83160     543286     84378        SET OF 3 HEART COOKIE CUTTERS        12  2011-02-07       1.25     14952.0  United Kingdom
384283    570116     22303             COFFEE MUG APPLES DESIGN         2  2011-10-07       2.55     12856.0  United Kingdom
436426    574138     23084                   RABBIT NIGHT LIGHT        96  2011-11-03       1.79     12415.0       Australia
253810    559174     35646        VINTAGE BEAD PINK EVENING BAG        24  2011-07-07       1.25     18263.0  United Kingdom
28921     538664     22173   METAL 4 HOOK HANGER FRENCH CHATEAU         8  2010-12-13       2.95     14733.0  United Kingdom
45827     540288     22928      YELLOW GIANT GARDEN THERMOMETER         1  2011-01-06       5.95     17634.0  United Kingdom
329567    565847     22846           BREAD BIN DINER STYLE RED          1  2011-09-07      33.29         NaN  United Kingdom
152572   C549666     22063  CERAMIC BOWL WITH STRAWBERRY DESIGN        -1  2011-04-11       2.95     13113.0  United Kingdom
116666    546275    82494L          WOODEN FRAME ANTIQUE WHITE          6  2011-03-10       2.95     13126.0  United Kingdom
31955     539034     22746        POPPY'S PLAYHOUSE LIVINGROOM          4  2010-12-15       2.10     16161.0  United Kingdom
314677    564648     23204          CHARLOTTE BAG APPLES DESIGN        10  2011-08-26       0.85     15641.0  United Kingdom
223391    556483     22626                 BLACK KITCHEN SCALES         1  2011-06-12       8.50     14823.0  United Kingdom
86801     543600     22646   CERAMIC STRAWBERRY CAKE MONEY BANK         4  2011-02-10       1.45     18168.0  United Kingdom
240665    558107     22381               TOY TIDY PINK POLKADOT         4  2011-06-26       2.10     17811.0  United Kingdom
152182    549577     21121    SET/10 RED POLKADOT PARTY CANDLES        24  2011-04-11       1.25     14286.0  United Kingdom
307572    563924     21293        MIRRORED DOVE WALL DECORATION         1  2011-08-21       2.10     15984.0  United Kingdom
144823    548808     22670            FRENCH WC SIGN BLUE METAL         2  2011-04-04       1.25     17774.0  United Kingdom
514174    579679     22113          GREY HEART HOT WATER BOTTLE         2  2011-11-30       4.25     13755.0  United Kingdom
443343    574700     22121           NOEL WOODEN BLOCK LETTERS          2  2011-11-06       5.95     16033.0  United Kingdom
18245  

In [41]:
# display(retail_data.describe())
# display(retail_data.info())
retail_data.isnull().sum()
# display(retail_data.nunique())

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerId     135080
Country             0
dtype: int64

| Column      | Unique Values | Null Values |
|:-------------:|:---------------:|:-------------:|
| InvoiceNo   | 25900         | 0           |
| StockCode   | 4070          | 0           |
| Description | 4223          | 1454        |
| Quantity    | 722           | 0           |
| InvoiceDate | 23260         | 0           |
| UnitPrice   | 1630          | 0           |
| CustomerID  | 4372          | 135080      |
| Country     | 38            | 0           |

This dataset contains 541,909 rows.

It's alarming that 135,080 (25%) of those rows contain null `CustomerID` values. This is a piece of important information to provide back to the store owner.  Additionally, 1,454 rows are missing a `Description` but, while that's a problem, the corresponding StockCode doesn't seem to be missing.  `Description` would be helpful to have from a readability standpoint but probably isn't needed for EDA.

I am choosing to keep the row with missing `CustomerID` rows for now because I may be able to do analysis on the data set (E.g., product analysis, time analysis, geospacial analysis) that won't require the data.

# Task 2: Clean the data


To Do Ideas:
1. Identify blank `Descriptions` and populate the correct value based on StockCode (Code of the Product), if possible.
2. Identify blank `CustomerID` and populate the correct valuea based on InvoiceNo, since multiple customers shouldn't be on the same Invoice?
   

## Fix missing Description values

Possible process:

- Create a list of `StockCode` and the most frequent corresponding `Description`. E.g. for StockCode "10080", there are three unique `Description` values: "check" (1), "GROOVY CACTUS INFLATABLE" (22), and "blank" (1). The most common `Description` is likely the valid value.
- Replace the existing value with the most-popular value
- I don't think it'll be easier to filter `StockCode` that have more than one unique `Description` but maybe.

### Count unique Descriptions per StockCode
I begin by counting the unique `Description` values per `StockCode`. Counts > 1 indiciate that the corresponding `StockCode` has more than one `Description` which, based on the description of the Dataset's columns, this shouldn't be possible.

In [42]:
unique_description_counts = retail_data.groupby('StockCode')['Description'].nunique().reset_index()
display(unique_description_counts[unique_description_counts['Description'] > 1])

StockCode  Description
1            10080            2
8            10133            2
26          15058A            2
28          15058C            2
31           16008            2
...            ...          ...
3972        90195A            2
4008        90210D            2
4043      DCGS0003            2
4050      DCGS0069            2
4065  gift_0001_20            2

[650 rows x 2 columns]

### Create a dictionary of Key:Value pairs using StockCode and most-popular Descriptions

The `most_common` function takes a single argument. If the Series is entirely composed of missing values, the function returns a None value. Otherwise, it returns the mode of each value in the `StockCode` column.

In [43]:
# Define a function to find the most common value or return None if the Series is empty.
def most_common(series):
    if series.dropna().empty:
        return None
    return series.mode().iloc[0]

# Creates a Lookup of all StockCode and hopefully-correct Descriptions. Group by 'StockCode' and apply the most_common function to 'Description'
stockcode_description_mode = retail_data.groupby('StockCode')['Description'].apply(most_common).reset_index()
display(stockcode_description_mode)

StockCode                         Description
0            10002         INFLATABLE POLITICAL GLOBE 
1            10080            GROOVY CACTUS INFLATABLE
2            10120                        DOGGY RUBBER
3           10123C               HEARTS WRAPPING TAPE 
4           10123G                                None
...            ...                                 ...
4065  gift_0001_20  Dotcomgiftshop Gift Voucher £20.00
4066  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00
4067  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00
4068  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00
4069             m                              Manual

[4070 rows x 2 columns]

The `stockcode_description_mode` DataFrame above contains 4070 rows, which matches the number UniqueValues identified in Task 1 for `StockCodes`. This suggests the code, so far, is correct.

I convert `stockcode_description_mode` to a dictionary `stockcode_description_dict` because I was getting errors in a step that used the `.replace()` and `.map()` functions.

In [44]:
# Convert DataFrame to dictionary
stockcode_description_dict = dict(zip(stockcode_description_mode['StockCode'], stockcode_description_mode['Description']))
# display(stockcode_description_dict)

I was not able to get the `.replace()` and `.map()` functions to "find and replace" the `StockCode` and `Description` values. As an alternative solution, I use the `.merge()` method to accomplish the same results. I clean up the results by dropping the original `Description` column (now `Description_x`), naming the new `Description_y` to `Description`, and reordering the columns.  

In [46]:
# Merge dictionary to retail_data
retail_data_new = pd.merge(retail_data, stockcode_description_mode, on='StockCode').drop(columns='Description_x').rename(columns={'Description_y':'Description'})

# Fix Column Order and Rename
retail_data_new = retail_data_new[['InvoiceNo', 'StockCode','Description', 'Quantity','InvoiceDate','UnitPrice','CustomerId','Country']]
display(retail_data_new)

InvoiceNo StockCode                          Description  Quantity InvoiceDate  UnitPrice  CustomerId         Country
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6  2010-12-01       2.55     17850.0  United Kingdom
1         536373    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6  2010-12-01       2.55     17850.0  United Kingdom
2         536375    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6  2010-12-01       2.55     17850.0  United Kingdom
3         536390    85123A   WHITE HANGING HEART T-LIGHT HOLDER        64  2010-12-01       2.55     17511.0  United Kingdom
4         536394    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32  2010-12-01       2.55     13408.0  United Kingdom
...          ...       ...                                  ...       ...         ...        ...         ...             ...
541904    579512     23617  SET 10 CARDS SWIRLY XMAS TREE 17104         2  2011-11-29       2.91         NaN  United Kingdom
541905    580691    90214U              LETTER U BLING KEY RING        12  2011-12-05       0.29     13790.0  United Kingdom
541906    581238    47591b             SCOTTIES CHILDRENS APRON         1  2011-12-08       4.13         NaN  United Kingdom
541907    581483     23843          PAPER CRAFT , LITTLE BIRDIE     80995  2011-12-09       2.08     16446.0  United Kingdom
541908   C581484     23843          PAPER CRAFT , LITTLE BIRDIE    -80995  2011-12-09       2.08     16446.0  United Kingdom

[541909 rows x 8 columns]

### Reviewing the fix

In [47]:
filt = retail_data_new['StockCode'] == 10080
retail_data_new.loc[filt]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerId, Country]
Index: []

`StockCode` 10080 had two "extra" `Description` values: "check" and a null value. Now, the "GROOVY CACTUS INFLATABLE" value is present for all 10080 rows!

When I re-run the same summary funcations on the new `retail_data_new` dataset, the `StockCode` numbers are significantly better.

In [48]:
# display(retail_data_new.describe())
# display(retail_data_new.info())
display(retail_data_new.isnull().sum())
# display(retail_data_new.nunique())

InvoiceNo           0
StockCode           0
Description       112
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerId     135080
Country             0
dtype: int64

In [49]:
filt_stockcode_description_mode = stockcode_description_mode[stockcode_description_mode['Description'].isna()]
filt_results = filt_stockcode_description_mode[['StockCode', 'Description']]
print(filt_results)

      StockCode Description
4        10123G        None
9         10134        None
46        16053        None
85       17011A        None
157       20689        None
...         ...         ...
4046   DCGS0057        None
4047  DCGS0066P        None
4052   DCGS0071        None
4053   DCGS0072        None
4055   DCGS0074        None

[112 rows x 2 columns]


I wanted to investigate the 112 `StockCode`s that were still blank. It turns out the corresponding `Description` values are all blank. Coincidentally, the row's `CustomerID` values are also blank. It would be my recommendation that these rows be dropped from the entire dataset.

## Fix missing CustomerID values 

This cannot be done the same way I fixed the Descriptions. When an InvoiceNo is missing a CustomerID, it's gone. For a given InvoiceID that has at least one missing CustomerID, there aren't other rows that have a CustomerID that have the same InvoiceNo.

The current dataset does not have enough information to execute a creative solution to impute missing data.

# Task 3: Basic Statistics

## Central Tendency
### Mean
**Mean (Average)** is the average value for a column. It is calculated as the sum of all values in the column divided by the number of data points.
- Mean of Quantity: mean_quantity = sum(Quantity) / count(Quantity)
- Mean of UnitPrice: mean_unit_price = sum(UnitPrice) / count(UnitPrice)

### Median
**Median** is the middle value in the sorted data. It's a measure of central tendency that is less affected by outliers.
- Median of Quantity: Sort the Quantity column and select the middle value.
- Median of UnitPrice: Sort the UnitPrice column and select the middle value.

### Mode
**Mode** is the most frequently occurring value in a column.
- Mode of Quantity: The value(s) that occur most frequently in the Quantity column.
- Mode of UnitPrice: The value(s) that occur most frequently in the UnitPrice column.

## Dispersion

### Range

The **range** gives you an idea of the spread of the data. It is calculated as the difference between the maximum and minimum values.
- Range of Quantity: range_quantity = max(Quantity) - min(Quantity)
- Range of UnitPrice: range_unit_price = max(UnitPrice) - min(UnitPrice)

### Variance

**Variance** measures the average squared difference of each data point from the mean. It quantifies how much the data points deviate from the mean.
- Variance of Quantity: `variance_quantity = sum((Quantity - mean_quantity)^2) / (count(Quantity) - 1)`
- Variance of UnitPrice: `variance_unit_price = sum((UnitPrice - mean_unit_price)^2) / (count(UnitPrice) - 1)`

### Standard Deviation

**Standard deviation** is the square root of the variance. It measures the average deviation from the mean and is often easier to interpret than variance.
- Standard deviation of Quantity: std_dev_quantity = sqrt(variance_quantity)
- Standard deviation of UnitPrice: std_dev_unit_price = sqrt(variance_unit_price)

### Interquartile Range
**Interquartile Range (IQR)** is the range between the first quartile (Q1) and the third quartile (Q3) of the data. It helps to understand the middle 50% of the data.
- Calculate Q1 and Q3, then IQR = Q3 - Q1 for Quantity and UnitPrice.